packages/core/src/external/commonwell/management/session.ts

```diff
@@ -0,0 +1,195 @@
+
+const cookiesToUse = [
+  "MP_RQ_COOKIE",
+  ".AspNet.Cookies",
+  "__RequestVerificationToken",
+  "SessionTimeoutRedirect",
+  "MPUserTimeZoneOffset",
+];
+
+/**
+ * Manages the session on the CommonWell management portal.
+ */
+export type SessionManagementConfig = {
+  username: string;
+  password: string;
+  cookieManager: CookieManager;
+  cwManagementApi: CommonWellManagementAPI;
+  /**
+   * The service to obtain a code to pass the login code challenge, if needed.
+   */
+  codeChallenge: CodeChallenge;
+  /**
+   * Indicates whether to run the browser in headless mode.
+   */
+  headless?: boolean;
+  debug?: typeof console.log;
+};
+
+export class SessionManagement {
+  private readonly cookieManager: CookieManager;
+  private readonly cwManagementApi: CommonWellManagementAPI;
+  private readonly codeChallenge: CodeChallenge;
+  private readonly headless: boolean;
+  private browser: Browser | undefined;
+  private context: BrowserContext | undefined;
+  private sessionBaseUrl: string;
+  private username: string;
+  private password: string;
+  private debug: typeof console.log;
+
+  constructor(params: SessionManagementConfig) {
+    this.cookieManager = params.cookieManager;
+    this.cwManagementApi = params.cwManagementApi;
+    this.codeChallenge = params.codeChallenge;
+    this.headless = params.headless == undefined ? true : params.headless;
+    this.sessionBaseUrl = this.cwManagementApi.getBaseUrl();
+    this.username = params.username;
+    this.password = params.password;
+    this.debug = params.debug ?? (() => {}); //eslint-disable-line @typescript-eslint/no-empty-function
+  }
+
+  async keepSessionActive(): Promise<void> {
+    const log = console.log;
+    try {
+      const isSessionValid = (await this.cwManagementApi.getMember()) != undefined;
+
+      if (isSessionValid) {
+        log(`Session is still valid, returning...`);
+        return;
+      }
+
+      log(`Session is invalid, generating new cookies...`);
+      const page = await this.login();
+
+      const cookies = await page.context().cookies();
+      const filteresCookies = cookies.filter(this.isEnabledCookie);
+      this.debug(`-----> cookiesString: ${cookiesToString(filteresCookies)}`);
+
+      log(`Got cookies, updating store...`);
+      await this.cookieManager.updateCookies(filteresCookies);
+
+      log(`Done.`);
+    } finally {
+      await this.logout();
+    }
+  }
+
+  /**
+   * Logs in on the CommonWell management portal.
+   *
+   * @returns the Playwright Page object pointing to CW's management portal Organization List page.
+   */
+  protected async login(): Promise<Page> {
+    const log = this.debug;
+    if (!this.browser || !this.browser.isConnected() || !this.context) {
+      this.browser = await firefox.launch({ headless: this.headless, slowMo: 50 });
+      this.context = await this.browser.newContext(devices["Desktop Firefox HiDPI"]);
+    }
+    const page = await this.context.newPage();
+
+    try {
+      log(`.1`);
```

Comment:
I left all these logs for now so we know exactly where it breaks, if that happens.